proof-systems/src/groth16/benches/gro16_bench.rs

```diff
@@ -0,0 +1,213 @@
+#[macro_use]
+extern crate criterion;
+
+use algebra::{
+    fields::bn_382::Fr,
+    curves::bn_382::Bn382,
+    UniformRand, PrimeField, Field,
+};
+use r1cs_core::{SynthesisError, ConstraintSynthesizer, ConstraintSystem, LinearCombination};
+use proof_systems::groth16::{generate_random_parameters, create_random_proof};
+
+use rand::SeedableRng;
+use rand_xorshift::XorShiftRng;
+
+use criterion::{BenchmarkId, BatchSize};
+use criterion::Criterion;
+
+use std::marker::PhantomData;
+
+pub struct Benchmark<F: PrimeField> {
+    inputs: Vec<Option<F>>,
+    num_constraints: usize,
+}
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for Benchmark<F> {
+    fn generate_constraints<CS: ConstraintSystem<F>>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(), SynthesisError> {
+        assert!(self.inputs.len() >= 2);
+        assert!(self.num_constraints >= self.inputs.len());
+
+        let mut variables: Vec<_> = Vec::with_capacity(self.inputs.len());
```

Comment:
Why allocating just inputs.len() many variables, as we know already how many variables the circuit has in advance?

Reply:
Pre-allocation size should be inputs.len() + num_constraints, fixed